crates/runtime/src/extension.rs

```diff
@@ -56,8 +68,23 @@ pub trait ExtensionRuntime: Send + Sync + 'static {
         _authorizer_id: AuthorizerId,
         _headers: http::HeaderMap,
     ) -> impl Future<Output = Result<(http::HeaderMap, Vec<u8>), ErrorResponse>> + Send;
+
+    fn authorize_query<'ctx>(
+        &'ctx self,
+        context: &'ctx Self::SharedContext,
+        extension_id: ExtensionId,
+        // (directive name, (definition, arguments))
+        elements: impl IntoIterator<
+            Item = (
+                &'ctx str,
+                impl IntoIterator<Item = DirectiveSite<'ctx, impl Anything<'ctx>>> + Send + 'ctx,
+            ),
+        > + Send
+        + 'ctx,
```

Comment:
This would look better as a generic parameter, so the long type signature is not polluting the param type defs.

Reply:
well ok, will change in another pr